lib/dynamic_type/test/dispatch.cpp

```diff
@@ -0,0 +1,166 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#include "dynamic_type/dynamic_type.h"
+
+#include <string>
+
+using namespace dynamic_type;
+
+class DynamicTypeTest : public ::testing::Test {};
+
+TEST_F(DynamicTypeTest, DispatchBasic) {
+  using IntOrStr = DynamicType<NoContainers, int64_t, std::string>;
+  auto concat = [](std::string a, auto x, auto y, std::string b) {
+    std::string result = a;
+    if constexpr (std::is_same_v<decltype(x), int64_t>) {
+      result += std::to_string(x);
+    } else if constexpr (std::is_same_v<decltype(x), std::string>) {
+      result += x;
+    }
+    if constexpr (std::is_same_v<decltype(y), int64_t>) {
+      result += std::to_string(y);
+    } else if constexpr (std::is_same_v<decltype(y), std::string>) {
+      result += y;
+    }
+    return result + b;
+  };
+
+  IntOrStr one(std::string("1"));
+  IntOrStr two(2);
+  EXPECT_TRUE(one.is<std::string>());
+  EXPECT_TRUE(two.is<int64_t>());
+
+  auto concated_result11 = IntOrStr::dispatch(concat, "0", one, one, "3");
+  static_assert(std::is_same_v<decltype(concated_result11), std::string>);
+  EXPECT_EQ(concated_result11, "0113");
+
+  auto concated_result12 = IntOrStr::dispatch(concat, "0", one, two, "3");
+  static_assert(std::is_same_v<decltype(concated_result12), std::string>);
+  EXPECT_EQ(concated_result12, "0123");
+
+  auto concated_result21 = IntOrStr::dispatch(concat, "0", two, one, "3");
+  static_assert(std::is_same_v<decltype(concated_result21), std::string>);
+  EXPECT_EQ(concated_result21, "0213");
+
+  auto concated_result22 = IntOrStr::dispatch(concat, "0", two, two, "3");
+  static_assert(std::is_same_v<decltype(concated_result22), std::string>);
+  EXPECT_EQ(concated_result22, "0223");
+}
+
+TEST_F(DynamicTypeTest, DispatchArgumentPerfectForwarding) {
+  using IntOrStr = DynamicType<NoContainers, int64_t, std::string>;
+  struct NonCopyable {
+    NonCopyable() = default;
+    NonCopyable(const NonCopyable&) = delete;
+  };
+  auto concat =
+      [](std::string& result, NonCopyable&& nc, auto x, auto y, std::string b) {
```

Comment:
What is the reason of having the `NonCopyable` parameter? It seems it's not used.

Reply:
It is just testing that the perfect forwarding is working correctly, because otherwise, we will likely try to create a copy of `nc`, which will lead to an error.